### PR TITLE
feat(edit-ema): #26988 Add layout editor 

### LINKED ISSUE #26988: Edit EMA: Render the template builder in `/layout`
### Parent
#26644 

### User Story
As a dotCMS user, I want to be able to update the layout of my page easily from the edit ema editor.

```[tasklist]
### Acceptance Criteria
- [x] The template builder must be rendered in the `/edit-ema/layout`
- [x] The template builder must work the exact same way as it works in the current edit content.
```

### Proposed Objective
Core Features

### Proposed Priority
Priority 3 - Average

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.:
N/A

### Assumptions & Initiation Needs
- The template builder new library is available.

### Quality Assurance Notes & Workarounds
N/A

### PR DESCRIPTION
**Pull request by zJaaal**
## Changes Proposed

- Move `DotRouterService` and dependencies to `libs`
- Move `CanDeactivateGuard` to `libs`
- Use this services to mimic the autosave from the old `edit-page`

## Screenshots

https://github.com/dotCMS/core/assets/63567962/75e9d007-02c4-4353-96f8-e234e5c01df2

### CODE REVIEW

File: core-web/libs/portlets/dot-ema/src/lib/services/dot-page-api.service.ts [hunk 1]
```diff
@@ -18,6 +24,9 @@ export interface DotPageApiResponse {
         language: DotLanguage;
         persona?: DotPersona;
     };
+    layout: DotLayout;
+    template: DotTemplate;
+    containers: DotPageContainerStructure;
```
Comment (fmontes) [inline core-web/libs/portlets/dot-ema/src/lib/services/dot-page-api.service.ts:29]: Why don't we do a `Pick` here of what we need? is just an idea so we can easily mock stuff around, well if that helps.

**Review by fmontes — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/dot-ema/src/lib/edit-ema-layout/edit-ema-layout.component.ts [hunk 2]
```diff
@@ -1,12 +1,213 @@
...
+            this.pageID = properties.pageID;
+
+            return {
+                ...properties,
+                containersMap: this.mapContainers(properties.containers)
+            };
+        })
+    );
+
+    private pageID: string;
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/dot-ema/src/lib/edit-ema-layout/edit-ema-layout.component.ts:58]: I saw it. I forgot to change it LOL

File: core-web/libs/portlets/dot-ema/jest.config.ts [hunk 3]
```diff
@@ -13,7 +13,7 @@ export default {
     transform: {
         '^.+\\.(ts|mjs|js|html)$': 'jest-preset-angular'
     },
-    transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'],
+    transformIgnorePatterns: ['<rootDir>/node_modules/(?!.*\\.mjs$)'],
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/dot-ema/jest.config.ts:16]: Yes! Gridstack is not supported by Jest, so we add this to ignore the whole node modules. You did it in Template Builder for the same reason. Here is the [docs](https://jestjs.io/docs/code-transformation)

File: core-web/apps/dotcms-ui/.storybook/tsconfig.json [hunk 4]
```diff
@@ -10,6 +10,8 @@
         "../../../**/block-editor/**/src/lib/**/*.stories.ts",
         "../../../**/edit-content/**/src/lib/**/*.stories.ts",
         "../../../**/ui/**/src/lib/**/*.stories.ts",
-        "../../../**/portlets/**/src/lib/**/*.stories.ts"
+        "../../../**/portlets/**/src/lib/**/*.stories.ts",
+        "../../../libs/data-access/src/lib/dot-router",
+        "../../../libs/dotcms-models/src/lib/navigation"
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/apps/dotcms-ui/.storybook/tsconfig.json:15]: I didn't explicitly changed this. I'll discard this.

File: core-web/libs/portlets/dot-ema/src/lib/edit-ema-editor/edit-ema-editor.component.spec.ts
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/dot-ema/src/lib/edit-ema-editor/edit-ema-editor.component.spec.ts:1]: Yes. Already fixed

File: core-web/libs/portlets/dot-ema/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts [hunk 5]
```diff
@@ -83,57 +85,27 @@ describe('EditEmaStore', () => {
...
-                done();
-            });
-        });
-
         it('should update editFrameURL', (done) => {
             spectator.service.setDialogIframeURL('test-url');
 
             spectator.service.state$.subscribe((state) => {
-                expect(state).toEqual({
+                expect(state as unknown).toEqual({
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/dot-ema/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts]: Tbh I did this to clean the logs to write the tests of layout. Indeed we need to fix this, thanks for the reminding me

File: core-web/libs/portlets/dot-ema/src/lib/edit-ema-layout/edit-ema-layout.component.spec.ts [hunk 6]
```diff
@@ -1,22 +1,183 @@
...
+        it('should forbid navigation', () => {
+            const spy = jest.spyOn(dotRouter, 'forbidRouteDeactivation');
+
+            templateBuilder.templateChange.emit();
+
+            expect(spy).toHaveBeenCalled();
+        });
+
+        it('should trigger an update template', () => {
+            const updateTemplate = jest.spyOn(component.updateTemplate, 'next');
```
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/dot-ema/src/lib/edit-ema-layout/edit-ema-layout.component.spec.ts:124]: LOL

File: core-web/libs/portlets/dot-ema/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (zJaaal) [reply core-web/libs/portlets/dot-ema/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts]: Fixed!

**Review by fmontes — APPROVED** [review]

**Review by KevinDavilaDotCMS — APPROVED** [review]

**Pull request merged**